internal/overlord/servstate/handlers.go

```diff
@@ -51,183 +54,528 @@ var (
 		}
-		_ = logBuffer.Close()
+		_ = s.logs.Close()
 		return fmt.Errorf("cannot start service: %v", err)
 	}
+	startTime, _ := s.config.ParseStartTime() // ignore error; it's already been validated
+	s.resetTimer = time.AfterFunc(startTime, func() { logError(s.startTimeElapsed()) })
 
-	active := &activeService{
-		originalPlan: service.Copy(),
-		cmd:          cmd,
-		done:         make(chan struct{}),
-		logBuffer:    logBuffer,
-	}
-	m.services[req.Name] = active
+	// Start a goroutine to wait for the process to finish.
+	done := make(chan struct{})
 	go func() {
-		active.err = cmd.Wait()
-		_ = active.logBuffer.Close()
-		close(active.done)
+		waitErr := s.cmd.Wait()
+		close(done)
+		err := s.exited(waitErr)
+		if err != nil {
+			logger.Noticef("Cannot execute exited action: %v", err)
+		}
 	}()
-	if m.serviceOutput != nil {
+
+	// Start a goroutine to read from the service's log buffer and copy to the output.
+	if s.manager.serviceOutput != nil {
 		go func() {
 			defer outputIterator.Close()
-			for outputIterator.Next(active.done) {
-				_, err := io.Copy(m.serviceOutput, outputIterator)
+			for outputIterator.Next(done) {
+				_, err := io.Copy(s.manager.serviceOutput, outputIterator)
 				if err != nil {
-					logger.Noticef("service %q log write failed: %v", req.Name, err)
+					logger.Noticef("service %q log write failed: %v", s.config.Name, err)
 				}
 			}
 		}()
 	}
 
-	releasePlan()
+	return nil
+}
 
-	okay := time.After(okayWait)
-	select {
-	case <-okay:
+// okayWaitElapsed is called when the okay-wait timer has elapsed (and the
+// service is considered running successfully).
+func (s *service) okayWaitElapsed() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	switch s.state {
+	case stateStarting:
+		s.started <- nil // still running fine after short duration, no error
+		s.transition(stateRunning)
+
+	default:
+		// Ignore if timer elapsed in any other state.
 		return nil
-	case <-active.done:
-		releasePlan, err := m.acquirePlan()
-		if err == nil {
-			if m.services[req.Name].cmd == cmd {
-				delete(m.services, req.Name)
+	}
+	return nil
+}
+
+// exited is called when the service's process exits.
+func (s *service) exited(waitErr error) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.resetTimer != nil {
+		s.resetTimer.Stop()
+	}
+
+	switch s.state {
+	case stateStarting:
+		code := exitCode(s.cmd)
+		s.started <- &code
+
+	case stateRunning:
+		logger.Noticef("Service %q stopped unexpectedly with code %d", s.config.Name, exitCode(s.cmd))
+		action, onType := getAction(s.config, waitErr == nil)
+		switch action {
+		case plan.ActionLog:
+			// Log has already been written above, no further log is necessary.
+			logger.Debugf("Service %q %s action is %q, transitioning to stopped state", s.config.Name, onType, action)
+			s.transition(stateStopped)
+
+		case plan.ActionExitPebble:
+			logger.Noticef("Service %q %s action is %q, triggering server exit", s.config.Name, onType, action)
```

Comment:
We need to review this name before it's final. I'm hoping we can find a term that implies termination clearly (and in contrast with the service termination) without having the self-reference to the project name.

Reply:
Done. Per discussion the other night, I've updated this to just "exit" (and "log" to "ignore").